app/models/assignment.rb

```diff
@@ -1247,29 +1247,122 @@ def to_json(options = {})
     self.assignment_properties.attributes.merge(self.attributes).symbolize_keys.to_json(options)
   end
 
+  # Returns an assignment's relevant properties for uploading/downloading an assignment's configuration as a hash
+  def assignment_properties_config
```

Comment:
I think this method is much more complex than it needs to be and is going to be a huge pain to maintain. Every time we change an attribute for an assignment object, we are going to have to change this method as well and I really would like to avoid that.

Here are a few suggestions:

- use `self.attributes` instead of repeating `self.send`
- just get all the attributes for each type of assignment. If they're not relevant for that assignment type then they'll have a nil/default value anyway.
- exclude only the attributes that are specific to this course (id, created_at, updated_at, all foreign keys, etc.). By excluding instead of including, we run less of a risk that changes will break this (if we add something new it will be included by default, if we remove something it won't be included anyway)

For example:

```ruby
def assignment_properties_config
    exclude = %w[id created_at updated_at outstanding_remark_request_count]
    properties = self.attributes.except(*exclude).reject { |attr| attr.end_with?('_id') }
    properties['assignment_properties_attributes'] = self.assignment_properties.attributes.except(*exclude).reject { |attr| attr.end_with?('_id') }
    properties['assignment_files_attributes'] = self.assignment_files.pluck_to_hash(*(AssignmentFile.column_names - exclude).reject { |attr| attr.end_with?('_id') })
    properties['submission_rule_attributes'] = self.submission_rule.attributes.except(*exclude).reject { |attr| attr.end_with?('_id') }
    properties['submission_rule_attributes']['periods_attributes'] = self.submission_rule.periods.pluck_to_hash(*(Period.column_names - exclude).reject { |attr| attr.end_with?('_id') })
end
```